File: 0007-reverse-integer/0007-reverse-integer.rb
def reverse(x)
    int_max = 2**31 - 1
    int_min = -2**31

    # Handle negative numbers
    if x < 0
        sign = -1
        x = x.abs
    else
        sign = 1
    end

    reversed_x = 0

    while x != 0
        # Pop the last digit
        pop = x % 10
        x = x / 10

        # Check for overflow
        if reversed_x > (int_max - pop) / 10
            return 0
        end

        reversed_x = reversed_x * 10 + pop
    end

    return sign * reversed_x
end

File: 0007-reverse-integer/0007-reverse-integer.js
/**
 * @param {number} x
 * @return {number}
 */
var reverse = function(x) {
    const INT_MIN = -2147483648; // -2^31
    const INT_MAX = 2147483647;  // 2^31 - 1

    let sign = Math.sign(x); // Get the sign of x
    let absReversed = parseInt(Math.abs(x).toString().split('').reverse().join('')); // Reverse the digits

    // Reapply the sign to the reversed number
    let result = sign * absReversed;

    // Check for 32-bit signed integer overflow
    if (result < INT_MIN || result > INT_MAX) {
        return 0;
    }

    return result;
};

